### PR TITLE
ar_tx_db overhaul

### PR DESCRIPTION
**Pull request by martin-torhage**

### CODE REVIEW

File: src/ar_tx.erl [hunk 1]
```diff
@@ -78,7 +78,12 @@ sign(TX, PrivKey, PubKey) ->
...
 -ifdef(DEBUG).
 verify(#tx { signature = <<>> }, _, _) -> true;
-verify(TX, Diff, WalletList) ->
+verify(TX, Diff, WalletList) -> do_verify(TX, Diff, WalletList).
+-else.
+verify(TX, Diff, WalletList) -> do_verify(TX, Diff, WalletList).
+-endif.
+
+do_verify(TX, Diff, WalletList) ->
 	% ar:report(
```
Comment (ldmberman) [inline src/ar_tx.erl]: May be let's remove this comment as well?

File: src/ar_tx_db.erl [hunk 2]
```diff
@@ -22,33 +21,49 @@ start() ->
...
+	{Priv2, Pub2} = ar_wallet:new(),
+	[B0] = ar_weave:init([
+		{ar_wallet:to_address(Pub1), ?AR(10000), <<>>},
+		{ar_wallet:to_address(Pub2), ?AR(10000), <<>>}
+	]),
+	%% Test bad transaction
+	OrphanedTX1 = ar_tx:new(Pub1, ?AR(1), ?AR(5000), <<>>),
+	BadTX = OrphanedTX1#tx { owner = Pub1, signature = <<"BAD">> },
+	?assert(not ar_tx:verify(BadTX, 8, B0#block.wallet_list)),
 	timer:sleep(500),
```
Comment (ldmberman) [inline src/ar_tx_db.erl]: Is `timer.sleep` here on purpose?

File: src/ar_tx_db.erl [hunk 3]
```diff
@@ -22,33 +21,49 @@ start() ->
...
-    ["tx_too_cheap ","tx_fields_too_large ","tag_field_illegally_specified ","last_tx_not_valid "] = get(TX#tx.id),
-    ar_tx:verify(SignedTX, 8, B0#block.wallet_list).
+	Expected = {ok, ["same_owner_as_target", "tx_id_not_valid", "tx_signature_not_valid"]},
+	?assertEqual(Expected, get_error_codes(BadTX#tx.id)),
+	%% Test good transaction
+	OrphanedTX2 = ar_tx:new(Pub1, ?AR(1), ?AR(5000), <<>>),
+	SignedTX = ar_tx:sign(OrphanedTX2, Priv2, Pub2),
+	?assert(ar_tx:verify(SignedTX, 8, B0#block.wallet_list)),
+	clear_error_codes(BadTX#tx.id),
+	clear_error_codes(SignedTX#tx.id),
```
Comment (ldmberman) [inline src/ar_tx_db.erl]: Makse sense to add an additional assertion to assert `clear_error_codes` actually cleans it up.

File: src/ar_key_db.erl [hunk 4]
```diff
@@ -1,13 +1,12 @@
...
 -include_lib("eunit/include/eunit.hrl").
 -include("ar.hrl").
-%%% Defines a small in-memory metadata table for Arweave nodes.
-%%% Typically used to store small peices of globally useful information
-%%% (for example: the port number used by the node).
+%%% Database for storing decryption keys from encrypted recall blocks. The
+%%% entries has a TTL. The DB is a singleton.
 
-%% @doc Initialise the metadata storage service.
+%% @doc Create the DB. This will fail if a DB already exists.
```
Comment (ldmberman) [inline src/ar_key_db.erl]: Not a big deal, but it should be `a DB` the first time and `the DB` the second, same below.

File: src/ar_tx_db.erl (hunk 2 above)
Comment (martin-torhage) [reply src/ar_tx_db.erl]: It's an old sleep but it doesn't make sense, so I successfully removed it now :)

File: src/ar_tx_db.erl (hunk 3 above)
Comment (martin-torhage) [reply src/ar_tx_db.erl]: `ets:delete/2` always return `true`, so there is nothing to test.

File: src/ar_tx.erl (hunk 1 above)
Comment (martin-torhage) [reply src/ar_tx.erl]: ✅

File: src/ar_key_db.erl (hunk 4 above)
Comment (martin-torhage) [reply src/ar_key_db.erl]: ✅

**Comment by martin-torhage** [comment]
@ldmberman I handled all your feedback. Please have a look again and see my reply.

File: src/ar_tx_db.erl (hunk 3 above)
Comment (ldmberman) [reply src/ar_tx_db.erl]: You can call `get_error_codes` after `clear_error_codes` and observe the absence of some previously added entries.

Comment (martin-torhage) [reply src/ar_tx_db.erl]: How about now?

**Pull request merged**